source/extensions/http/header_validators/envoy_default/http1_header_validator.cc

```diff
@@ -0,0 +1,359 @@
+      reject_details = UhvResponseCodeDetail::get().EmptyHeaderName;
+    } else {
+      auto entry_result = validateRequestHeaderEntry(header_name, header_value);
+      if (!entry_result) {
+        reject_details = static_cast<std::string>(entry_result.details());
+      }
+    }
+
+    return reject_details.empty() ? ::Envoy::Http::HeaderMap::Iterate::Continue
+                                  : ::Envoy::Http::HeaderMap::Iterate::Break;
+  });
+
+  return reject_details.empty()
+             ? RequestHeaderMapValidationResult::success()
+             : RequestHeaderMapValidationResult(RejectOrRedirectAction::Reject, reject_details);
+}
+
+::Envoy::Http::HeaderValidator::ResponseHeaderMapValidationResult
+Http1HeaderValidator::validateResponseHeaderMap(::Envoy::Http::ResponseHeaderMap& header_map) {
+  //
+  // Step 1: verify that required pseudo headers are present
+  //
+  // For HTTP/1.1 responses, RFC 7230 states that only the :status
+  // header is required: https://datatracker.ietf.org/doc/html/rfc7230#section-3.1.2
+  //
+  // status-line = HTTP-version SP status-code SP reason-phrase CRLF
+  //
+  if (header_map.getStatusValue().empty()) {
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidStatus};
+  }
+
+  //
+  // Step 2: Verify each response header
+  //
+  std::string reject_details;
+  header_map.iterate([this, &reject_details](const ::Envoy::Http::HeaderEntry& header_entry)
+                         -> ::Envoy::Http::HeaderMap::Iterate {
+    const auto& header_name = header_entry.key();
+    const auto& header_value = header_entry.value();
+    const auto& string_header_name = header_name.getStringView();
+
+    if (string_header_name.empty()) {
+      reject_details = UhvResponseCodeDetail::get().EmptyHeaderName;
+    } else {
+      auto entry_result = validateResponseHeaderEntry(header_name, header_value);
+      if (!entry_result) {
+        reject_details = static_cast<std::string>(entry_result.details());
+      }
+    }
+
+    return reject_details.empty() ? ::Envoy::Http::HeaderMap::Iterate::Continue
+                                  : ::Envoy::Http::HeaderMap::Iterate::Break;
+  });
+
+  return reject_details.empty()
+             ? ResponseHeaderMapValidationResult::success()
+             : ResponseHeaderMapValidationResult(RejectAction::Reject, reject_details);
+}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+Http1HeaderValidator::validateTransferEncodingHeader(const HeaderString& value) {
+  //
+  // HTTP/1.1 states that requests with an unrecognized transfer encoding should
+  // be rejected, from RFC 7230, https://tools.ietf.org/html/rfc7230#section-3.3.1:
+  //
+  // A server that receives a request message with a transfer coding it does not
+  // understand SHOULD respond with 501 (Not Implemented).
+  //
+  const auto& encoding = value.getStringView();
+  if (!absl::EqualsIgnoreCase(encoding, header_values_.TransferEncodingValues.Chunked)) {
```

Comment:
I think this is too restrictive. Envoy can accept compressed payloads using extensions: https://github.com/envoyproxy/envoy/blob/835dc5021aead1989a6e72ff165db91b58b127ff/source/common/http/headers.h#L327

I think we should just check if it is a list of `tokens`.